finance_dl/radius.py

```diff
@@ -0,0 +1,274 @@
+
+
+def check_url(url):
+    result = urllib.parse.urlparse(url)
+    if result.scheme != 'https' or not re.fullmatch(netloc_re, result.netloc):
+        raise RuntimeError('Reached invalid URL: %r' % url)
+
+
+standard_date_format = '%m/%d/%Y'
+
+class Scraper(scrape_lib.Scraper):
+    def __init__(self, credentials, output_directory,
+                 earliest_history_date=None, max_history_days=30,
+                 **kwargs):
+        """
+        @param earliest_history_date: Earliest UTC date for which to retrieve
+            transactions and balance information.
+
+        @param max_history_days: Number of days of history to retrieve, starting
+            from the previous UTC day, if `earliest_history_date` is not
+            specified.
+        """
+        super().__init__(**kwargs)
+        self.credentials = credentials
+        self.output_directory = output_directory
+        if not os.path.exists(self.output_directory):
+            os.makedirs(self.output_directory)
+        self.latest_history_date = (
+            datetime.datetime.now() - datetime.timedelta(days=1)).astimezone(
+                datetime.timezone.utc).date()
+        if earliest_history_date is None:
+            self.earliest_history_date = self.latest_history_date - datetime.timedelta(
+                days=max_history_days)
+        else:
+            self.earliest_history_date = dateutil.parser.parse(
+                earliest_history_date).date()
+        self.logged_in = False
+        if 'account_name' in kwargs.keys():
+            logger.info('Account name given: {}'.format(kwargs['account_name']))
+            self.account_name = kwargs['account_name']
+
+        print(kwargs)
+
+    def check_after_wait(self):
+        check_url(self.driver.current_url)
+
+    def login(self):
+        if self.logged_in:
+            return
+        logger.info('Initiating log in')
+        url = "https://banking.radiusbank.com/login"
+        self.driver.get(url)
+
+        # Click on that stupid button first
+        try:
+            continue_btn = self.wait_and_return(self.find_continue_button)
+            with self.wait_for_page_load():
+                continue_btn.click()
+        except:
+            pass
+
+        (username, password), = self.wait_and_return(
+            self.find_username_and_password_in_any_frame)
+        logger.info('Entering username and password')
+        # In case of autofill
+        for i in range(50):
+            username.send_keys(Keys.BACK_SPACE)
+
+        username.send_keys(self.credentials['username'])
+        password.send_keys(self.credentials['password'])
+        with self.wait_for_page_load(timeout = 30):
+            password.send_keys(Keys.ENTER)
+        logger.info('Logged in')
+        self.logged_in = True
+
+    def find_continue_btn(self):
+        for frame in self.for_each_frame():
+            try:
+                return self.driver.find_element_by_xpath('//button[@class="btn bg--aqua"]')
```

Comment:
If possible you may want to use different criteria --- as this looks like it may be brittle to changes in the website.

Reply:
Agree on brittleness. I haven't spent much time staring at HTML in general, so I would appreciate suggestions of the kinds of criteria that are less likely to change.

Reply:
Text or attributes that more directly relate to the function rather than the styling tend to be more reliable.  Sometimes you can bypass the HTML altogether and just look at the network requests being made (use network tab of browser developer tools).  If you can just generate the network requests (i.e. the URL and any necessary parameters / post data) directly, that tends to be easiest, fastest and most reliable.  Sometimes that is difficult though, in which case manipulating the UI directly is the way to go..

Reply:
Turns out that button check is unnecessary.